### PR TITLE
SDF formatting, review and validator

### PR DESCRIPTION
**Pull request by TSC21**
This plugin cleans up a bit the SDF files we have, based on the output of `xmllint` and using the supported and official SDF schema validator (http://sdformat.org/schemas/root.xsd). It also includes a script for doing this validation for both `model` and `world` SDF's, and should be used on CI for verifying that the modified SDFs don't have problems and are according to the approved schema.
The output of running `bash validate_sdf.bash` with all the SDF valid should be something like:
```bash
../models/shelves_high/model.sdf validates
../models/uneven_ground/model.sdf validates
../models/small_box/model.sdf validates
../models/delta_wing/delta_wing.sdf fails to validate
../models/flow_cam/model.sdf validates
../models/ramp/model.sdf validates
../models/big_box4/model.sdf validates
../models/solo/solo.sdf validates
../models/asphalt_plane/model-1_4.sdf validates
../models/asphalt_plane/model.sdf validates
../models/hippocampus/hippocampus.sdf validates
../models/iris_rplidar/iris_rplidar.sdf validates
../models/iris/iris.sdf validates
../models/big_box3/model.sdf validates
../models/shelves_high2_no_collision/model.sdf validates
../models/BoxesLargeOnPallet_3/model.sdf validates
../models/shelves_high2/model.sdf validates
../models/tailsitter/tailsitter.sdf validates
../models/big_box2/model.sdf validates
../models/lidar/model.sdf validates
../models/ground_plane/model.sdf validates
../models/plane/plane.sdf validates
../models/standard_vtol/standard_vtol.sdf validates
../models/Box/model.sdf validates
../models/polaris_ranger_ev/model-1_4.sdf validates
../models/polaris_ranger_ev/model.sdf validates
../models/polaris_ranger_ev/model-1_3.sdf validates
../models/sun/model.sdf validates
../models/rover/model.sdf validates
../models/rplidar/model.sdf validates
../models/BoxesLargeOnPallet/model.sdf validates
../models/BoxesLargeOnPallet_2/model.sdf validates
../models/iris_opt_flow/iris_opt_flow.sdf validates
../models/sonar/model.sdf validates
../models/big_box/model.sdf validates
../models/typhoon_h480/typhoon_h480.sdf validates
../models/irlock/model.sdf validates
../models/pallet_full/model.sdf validates
../models/iris_irlock/iris_irlock.sdf validates
../models/europallet/model.sdf validates
```
@dagar can we include this on CI?

### CODE REVIEW

**Comment by TSC21** [comment]
On validating the `worlds`, there is some issue with the `magnetic_field` element that I need to check how to solve yet. This applies to all `world` files on our repo.
```bash
../worlds/iris.world:37: element magnetic_field: Schemas validity error : Element 'magnetic_field': [facet 'pattern'] The value '6e-06 2.3e-05 -4.2e-05' is not accepted by the pattern '(\s*(-|\+)?(\d+(\.\d*)?|\.\d+|\d+\.\d+[eE][-\+]?[0-9]+)\s+){2}((-|\+)?(\d+(\.\d*)?|\.\d+|\d+\.\d+[eE][-\+]?[0-9]+))\s*'.
../worlds/iris.world:37: element magnetic_field: Schemas validity error : Element 'magnetic_field': '6e-06 2.3e-05 -4.2e-05' is not a valid value of the atomic type 'vector3'.
../worlds/iris.world fails to validate
```
I guess we have to change from scientific notation to decimal notation. Will try.

**Comment by TSC21** [comment]
Changing to decimal notation did the trick:
```bash
../worlds/iris.world validates
../worlds/hippocampus.world validates
../worlds/delta_wing.world validates
../worlds/rover.world validates
../worlds/typhoon_h480.world validates
../worlds/standard_vtol.world validates
../worlds/iris_rplidar.world validates
../worlds/plane.world validates
../worlds/iris_opt_flow.world validates
../worlds/tailsitter.world validates
../worlds/solo.world validates
../worlds/uneven.world validates
../worlds/rubble.world validates
../worlds/empty.world validates
../worlds/iris_irlock.world validates
../worlds/warehouse.world validates
```

**Comment by TSC21** [comment]
Now the `validate_sdf.bash` script is capable of usage in CI. @dagar can you take a look and maybe suggest a way of integrating this in the current `travis.yml`? Thanks

**Comment by TSC21** [comment]
Here's an example of a passed validation (this a validation made by the script over the changes done on this PR):
```bash
Validating Gazebo worlds at /home/tsc21/PX4_dev/src/sitl_gazebo/models:
/home/tsc21/PX4_dev/src/sitl_gazebo/models/shelves_high/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/uneven_ground/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/small_box/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/flow_cam/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/ramp/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/big_box4/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/solo/solo.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/asphalt_plane/model-1_4.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/asphalt_plane/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/hippocampus/hippocampus.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/iris_rplidar/iris_rplidar.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/iris/iris.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/big_box3/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/shelves_high2_no_collision/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/BoxesLargeOnPallet_3/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/shelves_high2/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/tailsitter/tailsitter.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/big_box2/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/lidar/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/ground_plane/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/plane/plane.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/standard_vtol/standard_vtol.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/Box/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/polaris_ranger_ev/model-1_4.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/polaris_ranger_ev/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/polaris_ranger_ev/model-1_3.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/sun/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/rover/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/rplidar/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/BoxesLargeOnPallet/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/BoxesLargeOnPallet_2/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/iris_opt_flow/iris_opt_flow.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/sonar/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/big_box/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/typhoon_h480/typhoon_h480.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/irlock/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/pallet_full/model.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/iris_irlock/iris_irlock.sdf validates
/home/tsc21/PX4_dev/src/sitl_gazebo/models/europallet/model.sdf validates
Validating Gazebo worlds at /home/tsc21/PX4_dev/src/sitl_gazebo/worlds:
/home/tsc21/PX4_dev/src/sitl_gazebo/worlds/iris.world validates
/home/tsc21/PX4_dev/src/sitl_gazebo/worlds/hippocampus.world validates
/home/tsc21/PX4_dev/src/sitl_gazebo/worlds/delta_wing.world validates
/home/tsc21/PX4_dev/src/sitl_gazebo/worlds/rover.world validates
/home/tsc21/PX4_dev/src/sitl_gazebo/worlds/typhoon_h480.world validates
/home/tsc21/PX4_dev/src/sitl_gazebo/worlds/standard_vtol.world validates
/home/tsc21/PX4_dev/src/sitl_gazebo/worlds/iris_rplidar.world validates
/home/tsc21/PX4_dev/src/sitl_gazebo/worlds/plane.world validates
/home/tsc21/PX4_dev/src/sitl_gazebo/worlds/iris_opt_flow.world validates
/home/tsc21/PX4_dev/src/sitl_gazebo/worlds/tailsitter.world validates
/home/tsc21/PX4_dev/src/sitl_gazebo/worlds/solo.world validates
/home/tsc21/PX4_dev/src/sitl_gazebo/worlds/uneven.world validates
/home/tsc21/PX4_dev/src/sitl_gazebo/worlds/rubble.world validates
/home/tsc21/PX4_dev/src/sitl_gazebo/worlds/empty.world validates
/home/tsc21/PX4_dev/src/sitl_gazebo/worlds/iris_irlock.world validates
/home/tsc21/PX4_dev/src/sitl_gazebo/worlds/warehouse.world validates
All SDFs validated positively!
```

**Comment by dagar** [comment]
What about installing the deps directly on travis-ci and running the validation at the end of the regular build or even in another travis-ci thread. https://github.com/PX4/sitl_gazebo/blob/master/.travis.yml#L15

**Comment by LorenzMeier** [comment]
Yes we should do that. And I don’t like that it rejects scientific notation. That just seems wrong and will cause a lot of drama down the road. Is there a better written validator?

**Comment by TSC21** [comment]
@LorenzMeier this is one that I found to be the "official" schema. We can see how the mag is handled on the world and check if it actually has to be included as decimal. But this was the only solution I found based on the error above.
@dagar yes that needs to be done. I added also to the README so users can also use the functionallity locally in their machines, for example, to verify the changes on done on sdfs before commiting/pull requesting.

**Comment by TSC21** [comment]
@LorenzMeier I was able to fix the `magnetic_field` element in order to keep the scientific notation. We can see that it states `Element 'magnetic_field': [facet 'pattern'] The value '6e-6 2.3e-5 -4.2e-5' is not accepted by the pattern '(\s*(-|\+)?(\d+(\.\d*)?|\.\d+|\d+\.\d+[eE][-\+]?[0-9]+)\s+){2}((-|\+)?(\d+(\.\d*)?|\.\d+|\d+\.\d+[eE][-\+]?[0-9]+))\s*'.`. So if we change `6e-6 2.3e-5 -4.2e-5` to `6.0e-6 2.3e-5 -4.2e-5`, everything works now (we were just missing the decimal part of the x-axis).

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,6 @@
 Build/
 build/
+scritps/schemas
```
Comment (anassinator) [inline .gitignore]: typo?

**Comment by TSC21** [comment]
@LorenzMeier @dagar scripts updated. Now it takes around 10 seconds to download the schemas and validate the SDFs, while before it took 10 minutes to do the same, given that it would download the root and subsequent schemas for every file it was validating.

This way, when running `validate_sdf.bash` it downloads and creates a local copy of the `xsd` files, changes the respective URL's to point to the local files and then runs the validation based on the local `xsd` files, taking around 10 to 15 seconds to finish all the process (download, validate, delete schemas folder since not needed). If one wants to keep the schemas locally, it can also do it by modifying the script for not remove the folder.

So right now I consider this good to go and ready to be integrate on CI. Waiting your reviews.

File: .gitignore (hunk 1 above)
Comment (TSC21) [reply .gitignore]: yeah thanks

**Comment by TSC21** [comment]
Now integrated into CI and passing. @dagar seems like somehow the `libxml2-utils` wasn't really required as I'm still using a previous tag on the travis config.

**Review by dagar — APPROVED** [review]
Looks good!

**Pull request merged**